From bhutchings@solarflare.com  Thu Jun 23 20:54:32 2011
From: Ben Hutchings <bhutchings@solarflare.com>
Date: Thu, 9 Dec 2010 12:10:25 +0000
Subject: sfc: Report link-dwn while interface is down
Patch-mainline: Never
References: bnc#698572, FATE#311724

This should have the same effect in sfc as the upstream commits
e596e6e4d578f2639416e620d367a3af34814a40 (ethtool core) and
ed4ba4b5b96742d29225308ccccbdb1810b63064 (sfc).

Signed-off-by: Ben Hutchings <bhutchings@solarflare.com>
Acked-by: Jeff Mahoney <jeffm@suse.com>
---
 drivers/net/sfc/ethtool.c |    9 +--------
 1 files changed, 1 insertions(+), 8 deletions(-)

diff --git a/drivers/net/sfc/ethtool.c b/drivers/net/sfc/ethtool.c
index 5e50e57..0e8bb19 100644
--- a/drivers/net/sfc/ethtool.c
+++ b/drivers/net/sfc/ethtool.c
@@ -622,9 +622,7 @@ static int efx_ethtool_nway_reset(struct net_device *net_dev)
 
 static u32 efx_ethtool_get_link(struct net_device *net_dev)
 {
-	struct efx_nic *efx = netdev_priv(net_dev);
-
-	return efx->link_state.up;
+	return netif_running(net_dev) && netif_carrier_ok(net_dev);
 }
 
 static int efx_ethtool_get_coalesce(struct net_device *net_dev,

